README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

<img src="man/figures/sticker.png" align="right" width="15%" height="15%" />

## Overview

The `FreeHiCLite` package is designed for simulate Hi-C contact matrix.

The original package [FreeHi-C](https://github.com/yezhengSTAT/FreeHiC)
is short for **Fr**agment interactions **e**mpirical **e**stimation for
fast simulation of **Hi-C** data. It is a data-driven Hi-C data
simulator for simulating and augmenting Hi-C datasets. FreeHi-C employs
a non-parametric strategy for estimating an interaction distribution of
genome fragments and simulates Hi-C reads from interacting fragments.
Data from FreeHi-C exhibit higher fidelity to the biological Hi-C data.
FreeHi-C not only can be used to study and benchmark a wide range of
Hi-C analysis methods but also boosts power and enables false discovery
rate control for differential interaction detection algorithms through
data augmentation. Different from FreeHi-C (v1.0), a spike-in module is
added enabling the simulation of true differential chromatin
interactions.

FreeHi-C is designed for studies that are prone to simulate Hi-C
interactions from the real data and add deviations from the true ones.
Therefore, FreeHi-C requires real Hi-C sequencing data (FASTQ format) as
input along with user-defined simulation parameters. FreeHi-C will
eventually provide the simulated genomics contact counts in a sparse
matrix format (BED format) which is compatible with the standard input
of downstream Hi-C analysis.

## Installation

Install the development version using the **devtools** package:

``` r
devtools::install_github("banconzhou/FreeHiCLite")
```

## Usage

Load the package:

``` r
library(FreeHiCLite)
```

### Read hic data

The [`.hic`](https://github.com/aidenlab/juicer/wiki/Data) file is a
highly compressed binary file, which is developed in the [Aiden
Lab](http://aidenlab.org/). Which can be used in
[juicebox](https://aidenlab.org/juicebox/) for contact matrix
visualization.

The `.hic` file is formatted as [HiC
Format](https://github.com/aidenlab/Juicebox/blob/master/HiCFormatV8.md).
To program with `.hic` file, they provide
[straw](https://github.com/aidenlab/straw) and
[Dump](https://github.com/aidenlab/juicer/wiki/Data-Extraction) to
extract the information from the `.hic` file. The `readJuicer()` adopts
most from the C++ version of [straw](https://github.com/aidenlab/straw).

The `.hic` file only contains two units of resolution, and each unit
contains a fix set of resolutions.

1.  Base-pair-delimited resolutions (**BP**): 2.5M, 1M, 500K, 250K,
    100K, 50K, 25K, 10K, and 5K.
2.  Fragment-delimited resolutions (**FRAG**): 500f, 250f, 100f, 50f,
    20f, 5f, 2f,
1f.

<!-- end list -->

``` r
## Remote file location. The reomte file include downloading, it may take a while
remoteFilePath <- "https://hicfiles.s3.amazonaws.com/hiseq/gm12878/in-situ/combined.hic"
 
## Local file location
localFilePath <- system.file("extdata", "example.hic", package = "FreeHiCLite")
 
## Chromosomes needs to be extra
chromosomes <- c("chr1", "chr2")

## Pairs needs to be extra
pairs <- c("1_1", "1_2")
unit <- "BP"
resolution <- 500000L
```

#### Extract hic information

We can extract some basic information from a `.hic` file via:

``` r
juicerInfo <- readJuicerInformation(localFilePath, verbose = TRUE)
#> File: /Library/Frameworks/R.framework/Versions/4.0/Resources/library/FreeHiCLite/extdata/example.hic
#> GenomId:
#> Available pair: 1_1, 1_2, 1_3, 2_2, 2_3, 3_3.
#> Hi-C resolution:
#>     BP: 2500000, 500000, 5000.
#>     FRAG not available.
```

Besides, `juicerInfo()` also provides chromosome size information, if
the genomeID is not one of the genome they provided. [check
here](https://github.com/aidenlab/juicer/wiki/Pre#usage)

``` r
head(juicerInfo[["chromosomeSizes"]])
#>   chromosome      size
#> 1          1 249250621
#> 2         10 135534747
#> 3         11 135006516
#> 4         12 133851895
#> 5         13 115169878
#> 6         14 107349540
```

#### Read hic file

To read a remote `.hic` file. Provide a remote url to `readJuicer()`.
The full list of remote links can be found in
<http://aidenlab.org/data.html>.

``` r
## pass chrosomes into function, it will contains all the interaction pairs
datRemote <- readJuicer(file=remoteFilePath, chromosomes=chromosomes, pairs = NULL, unit=unit, resolution=resolution)
```

``` r
## pass chrosomes into function, it will contains all the interaction pairs
datLoc <- readJuicer(file=localFilePath, chromosomes=chromosomes, pairs = NULL, unit=unit, resolution=resolution)
str(datLoc)
#> List of 3
#>  $ contact    :List of 3
#>   ..$ 1_1: int [1:87158, 1:3] 500000 500000 1000000 500000 1000000 1500000 500000 1000000 1500000 2000000 ...
#>   .. ..- attr(*, "dimnames")=List of 2
#>   .. .. ..$ : NULL
#>   .. .. ..$ : chr [1:3] "x" "y" "counts"
#>   ..$ 1_2: int [1:69341, 1:3] 2500000 4000000 4500000 5000000 5500000 6000000 7000000 8500000 9500000 10000000 ...
#>   .. ..- attr(*, "dimnames")=List of 2
#>   .. .. ..$ : NULL
#>   .. .. ..$ : chr [1:3] "x" "y" "counts"
#>   ..$ 2_2: int [1:104782, 1:3] 0 0 500000 0 500000 1000000 0 500000 1000000 1500000 ...
#>   .. ..- attr(*, "dimnames")=List of 2
#>   .. .. ..$ : NULL
#>   .. .. ..$ : chr [1:3] "x" "y" "counts"
#>  $ information:List of 4
#>   ..$ genomeID       : chr "hg19"
#>   ..$ resolution     :List of 2
#>   .. ..$ BP  : int [1:3] 2500000 500000 5000
#>   .. ..$ FRAG: int(0) 
#>   ..$ pairs          : chr [1:6] "1_1" "1_2" "1_3" "2_2" ...
#>   ..$ chromosomeSizes:'data.frame':  26 obs. of  2 variables:
#>   .. ..$ chromosome: chr [1:26] "1" "10" "11" "12" ...
#>   .. ..$ size      : int [1:26] 249250621 135534747 135006516 133851895 115169878 107349540 102531392 90354753 81195210 78077248 ...
#>  $ settings   :List of 4
#>   ..$ unit       : chr "BP"
#>   ..$ resolution : int 500000
#>   ..$ chromosomes: chr [1:2] "chr1" "chr2"
#>   ..$ file       : chr "/Library/Frameworks/R.framework/Versions/4.0/Resources/library/FreeHiCLite/extdata/example.hic"
#>  - attr(*, "class")= chr [1:2] "juicer" "freehic"
```

#### Write to disk file

The [juicebox](https://aidenlab.org/juicebox/) also provide a function
[Pre](https://github.com/aidenlab/juicer/wiki/Pre) to generate `.hic`
file from different format. In `FreeHiCLite`, we provide a function
`writeJuicer()` to write the contact matrix into a [short with score
format](https://github.com/aidenlab/juicer/wiki/Pre#short-with-score-format).
You can use `convertJuicer()` to view the matrix, or directly use
`writeJuicer()`.

``` r
head(convertJuicer(datLoc[['contact']][[1]], "1", "1"))
#>   str1 chr1    pos1 frag1 str2 chr2    pos2 frag2 score
#> 1    0    1  500000     0    0    1  500000     1   384
#> 2    0    1  500000     0    0    1 1000000     1   231
#> 3    0    1 1000000     0    0    1 1000000     1  1272
#> 4    0    1  500000     0    0    1 1500000     1    47
#> 5    0    1 1000000     0    0    1 1500000     1   373
#> 6    0    1 1500000     0    0    1 1500000     1  1665
```

``` r
writeJuicer(datLoc, file, overwrite = TRUE)
```

### Add spikeIn

To add spikeIn into contact matrix. The `FreeHiCLite` provides
`FreeSpikeIn()` to add spikein.

``` r
kernelSmooth <- TRUE
bandwidth <- 50000L

## Create a random spikeIn

contact <- datLoc[['contact']][[1]]
Ns <- 0.1 * NROW(contact)
spikeIn <- contact[sample(1:NROW(contact), Ns), ]
spikeIn[,3] <- spikeIn[,3] * sample(seq(0, 10, 0.5), Ns, replace=TRUE) # 3rd is the counts

spikeInContact <- FreeSpikeIn(contact, spikeIn, kernelSmooth = kernelSmooth, bandwidth = bandwidth)
print(summary(contact[,3]))
#>    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#>    1.00    2.00    4.00   25.71    9.00 4545.00
print(summary(spikeInContact[,3]))
#>    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#>     1.0     2.0     5.0   149.1    14.0  4545.0
```

### Simulate from contact matrix

Here we use `FreeHiC()` function to perform
simulation.

``` r
simuContact <- FreeHiC(spikeInContact, seqDepth = 2 * sum(contact[,3]), resolution = 5000L)
print(summary(simuContact[,3]))
#>    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#>     1.0     2.0     6.0   163.7    16.0  4594.0
```

## References

1.  Ye Zheng and Sündüz Keleş. [“FreeHi-C simulates high-fidelity Hi-C
    data for benchmarking and data
    augmentation.”](https://www.nature.com/articles/s41592-019-0624-3)
    Nature Methods (2019).
2.  Neva C. Durand, Muhammad S. Shamim, Ido Machol, Suhas S. P. Rao,
    Miriam H. Huntley, Eric S. Lander, and Erez Lieberman Aiden.
    [“Juicer provides a one-click system for analyzing loop-resolution
    Hi-C
    experiments.”](https://www.cell.com/cell-systems/fulltext/S2405-4712\(16\)30219-8)
    Cell Systems 3(1), 2016.


NEWS.md
# FreeHiCLite 0.1.0

* Fix a major bug in windows. long size is 4 in windows. Change to long long. File type change to binary.
* Added vignettes about how to use `FreeHiC()` and `readJuicer()`
* Added a `NEWS.md` file to track changes to the package.


R/FreeHiC.R
#' @title 
#' 
#' FreeHiC simulator
#' 
#' @description 
#' 
#' This function performs FreeHiC based on contact matrix only.
#' 
#' @param contacts A \code{\link[base]{list}} of contacts matrix. The list element should named by chromosome pair.
#' \itemize{
#' \item{\code{\link[base]{matrix}}}{A matrix with columns (x, y, counts).}
#' \item{\code{\link[base]{data.frame}}}{A data.frame with columns (chr1, x, y, chr2, counts).}
#' \item{\code{\link[base]{list}}}{A list of matrix, where the matrix has columns (x, y, counts).}
#' }
#' where \code{x} stands for the first chromosome location, \code{y} stands for the second chromosome location and \code{counts} is the interaction counts.
#' @param seqDepth The desired sequence depth.
#' @param countScale The scale of counts. A number larger than 0. If both \code{seqDepth} and \code{countScale} are provided. Choose the larger one.
#' @param noiseRate The noise rate for contact matrix. 0 - 1 scale
#' @param neighborZeroRate The rate for neighborhood noise rate. 0 - 1 scale
#' @param resolution The resolution used in the contacts matrix. A positive number.
#' 
#' @return A list or matrix or data.frame with the same format as \code{contacts}
#' 
#' @examples 
#' 
#' library(FreeHiCLite)
#' 
#' N <- 2000
#' maxBinX <- maxBinY <- 2000000L
#' maxCounts <- 10
#' binX <- sample(1:maxBinX, N, replace=TRUE)
#' binY <- sample(1:maxBinY, N, replace=TRUE)
#' counts <- sample(1:10, N, replace=TRUE)
#' 
#' seqDepth <- 20000
#' countScale <- 0
#' noiseRate <- 0.1
#' neighborZeroRate <- 0
#' resolution <- 5000
#' 
#' ## Matrix version
#' ## matrix layout as x, y, counts
#' contacts <- matrix(0, N, 3)
#' contacts[,1] <- binX
#' contacts[,2] <- binY
#' contacts[,3] <- counts
#' 
#' res <- FreeHiC(contacts, seqDepth, countScale, noiseRate, neighborZeroRate, resolution)
#' head(res)
#' 
#' ## List version
#' N2 = 3000
#' binX2 <- sample(1:maxBinX, N2, replace=TRUE)
#' binY2 <- sample(1:maxBinY, N2, replace=TRUE)
#' counts2 <- sample(1:10, N2, replace=TRUE)
#' 
#' contacts2 <- matrix(0, N2, 3)
#' contacts2[,1] <- binX2
#' contacts2[,2] <- binY2
#' contacts2[,3] <- counts2
#' 
#' contactsMap <- list('1_1' = contacts, '2_2' = contacts2)
#' res <- FreeHiC(contactsMap, seqDepth, countScale, noiseRate, neighborZeroRate, resolution)
#' str(res)
#' 
#' ## Dataframe version
#' 
#' chr1 <- c(rep('1', N), rep('2', N2))
#' chr2 <- c(rep('1', N), rep('2', N2))
#' 
#' contactsAll <- rbind(contacts, contacts2)
#' 
#' contactsDf <- data.frame(chr1 = chr1, x = contactsAll[,1], 
#' chr2 = chr2, y = contactsAll[,2], counts = contactsAll[,3])
#' 
#' res <- FreeHiC(contactsDf, seqDepth, countScale, noiseRate, neighborZeroRate, resolution)
#' head(res)
#' 
#' @references 
#' Zheng, Ye, Keles, Sunduz FreeHi-C simulates high-fidelity Hi-C data for benchmarking and data augmentation. 
#' \emph{Nature Methods} \strong{17}, 37–40 (2020). \href{https://doi.org/10.1038/s41592-019-0624-3}{doi}
#' 
#' @importFrom methods is
#' 
#' @export
FreeHiC <- function(contacts, seqDepth = NULL, countScale = 1, noiseRate = 0, neighborZeroRate = 0, 
    resolution = 50000L) {
    
    stopifnot(is.null(seqDepth) | seqDepth >= 0)
    stopifnot(countScale > -1)
    stopifnot(noiseRate >= 0 & noiseRate < 1)
    stopifnot(neighborZeroRate >= 0 & neighborZeroRate < 1)
    stopifnot(resolution >= 0)
    
    if (methods::is(contacts, "list")) {
        ans <- .FreeHiCList(contactsMap = contacts, seqDepth = seqDepth, countScale = countScale, noiseRate = noiseRate, 
            neighborZeroRate = neighborZeroRate, resolution = resolution)
    } else if (methods::is(contacts, "data.frame")) {
        ans <- .FreeHiCDf(contactsDf = contacts, seqDepth = seqDepth, countScale = countScale, noiseRate = noiseRate, 
            neighborZeroRate = neighborZeroRate, resolution = resolution)
    } else if (methods::is(contacts, "matrix")) {
        ans <- .FreeHiCMatrix(contactsMatrix = contacts, seqDepth = seqDepth, countScale = countScale, 
            noiseRate = noiseRate, neighborZeroRate = neighborZeroRate, resolution = resolution)
    } else {
        stop("Not implemented. Check ?FreeHIC")
    }
    return(ans)
}

.FreeHiCList <- function(contactsMap, seqDepth, countScale, noiseRate, neighborZeroRate, resolution) {
    pairs <- names(contactsMap)
    stopifnot(!is.null(pairs))
    return(hicDataSimuList(contactRecords = contactsMap, names = pairs, resolution = resolution, sequenceDepth = seqDepth, 
        countScale = countScale, noiseRate = noiseRate, neighborZeroRate = neighborZeroRate))
}

.FreeHiCDf <- function(contactsDf, seqDepth, countScale, noiseRate, neighborZeroRate, resolution) {
    stopifnot(NCOL(contactsDf) == 5)
    contactsMap <- .dfToList(contactsDf)
    resList <- .FreeHiCList(contactsMap = contactsMap, seqDepth = seqDepth, countScale = countScale, 
        noiseRate = noiseRate, neighborZeroRate = neighborZeroRate, resolution = resolution)
    pairs <- names(resList)
    counts <- sapply(resList, NROW)
    chrs <- sapply(pairs, FUN = function(x) {
        strsplit(x, "_")
    })
    
    chr1 <- c()
    chr2 <- c()
    
    for (i in seq_along(counts)) {
        chr1 <- c(chr1, rep(chrs[[i]][1], counts[i]))
        chr2 <- c(chr2, rep(chrs[[i]][2], counts[i]))
    }
    
    mat <- do.call("rbind", resList)
    rm(resList)
    
    df <- data.frame(chr1 = chr1, x = mat[, 1], chr2 = chr2, y = mat[, 2], counts = mat[, 3])
    names(df) <- names(contactsDf)
    return(df)
}

.FreeHiCMatrix <- function(contactsMatrix, seqDepth, countScale, noiseRate, neighborZeroRate, resolution) {
    stopifnot(NCOL(contactsMatrix) == 3)
    return(hicDataSimuMatrix(contactRecords = contactsMatrix, resolution = resolution, sequenceDepth = seqDepth, 
        countScale = countScale, noiseRate = noiseRate, neighborZeroRate = neighborZeroRate))
}

#' @title
#' 
#' FreeHiC directly from hic file
#' 
#' @param file Filename can be a local path or remote path. 
#' The remote path full list can be obtained from \url{http://aidenlab.org/data.html}.
#' @param chromosomes A vector contains all the chromosomes. 
#' For example c('chr1', 'chr2'), the resulting contact matrixes will include all the pairs of interaction (chr1_chr1, chr1_chr2, chr2_chr2). 
#' @param pairs A vector contains all the pair. 
#' The pair take format as '1_1' or 'chr1_chr1', both means the contact between chromosome1 and chromosome1. If \code{pair} presents, \code{chromosomes} argument will be ignore.
#' @param unit Unit only supports c('BP', 'FRAG'). 
#' 'BP' means base-pair, and 'FRAG' means fragment.
#' @param resolution The resolution used in the contacts matrix. A positive number.
#' The resolution must be a value from following list.
#' \itemize{
#' \item{\strong{unit: BP}}
#' \itemize{
#' \item 2500000
#' \item 1000000
#' \item 500000
#' \item 250000
#' \item 100000
#' \item 50000
#' \item 25000
#' \item 10000
#' \item 5000
#' }
#' \item{\strong{unit: FRAG}}
#' \itemize{
#' \item 500
#' \item 250
#' \item 100
#' \item 50
#' \item 20
#' \item 5
#' \item 2
#' \item 1
#' }
#' }
#' @param seqDepth The desired sequence depth.
#' @param countScale The scale of counts. A number larger than 0. If both \code{seqDepth} and \code{countScale} are provided. Choose the larger one.
#' @param noiseRate The noise rate for contact matrix. 0 - 1 scale
#' @param neighborZeroRate The rate for neighborhood noise rate. 0 - 1 scale
#' @param verbose TRUE or FALSE. Whether print information.
#' 
#' @return A list contains simulated contact matrix and basic information of hic data
#' 
#' @examples
#' 
#' library(FreeHiCLite)
#' 
#' ## Local file location
#' localFilePath = system.file('extdata', 'example.hic', package = 'FreeHiCLite')
#' 
#' ## Chromosomes needs to be extracted
#' chromosomes = c('chr1', 'chr2')
#' 
#' ## Pairs needs to be extracted
#' pairs = c('1_1', '1_2')
#' unit = 'BP'
#' resolution = 500000L
#' 
#' seqDepth <- 20000
#' countScale <- 0
#' noiseRate <- 0.1
#' neighborZeroRate <- 0
#' 
#' 
#' ## pass chrosomes into function, it will contains all the interaction pairs
#' 
#' res <- FreeHiCJuicer(file=localFilePath, chromosomes=chromosomes, 
#' pairs = NULL, unit=unit, resolution=resolution,
#' seqDepth = seqDepth, countScale = countScale, 
#' noiseRate = noiseRate, neighborZeroRate = neighborZeroRate)
#' 
#' str(res)
#' 
#' 
#' @export
FreeHiCJuicer <- function(file, chromosomes = NULL, pairs = NULL, unit = c("BP", "FRAG"), seqDepth = NULL, 
    countScale = 1, noiseRate = 0, neighborZeroRate = 0, resolution = 50000L, verbose=TRUE) {
    
    defResolution <- list(BP = c(2500000, 1e+06, 5e+05, 250000, 1e+05, 50000, 25000, 10000, 5000), 
        FRAG = c(500, 250, 100, 50, 20, 5, 2, 1))
    
    isHttp <- startsWith(file, "http")
    if (!isHttp) {
        file = normalizePath(file)
        stopifnot(file.exists(file))
    }
    if(verbose) message("You are trying to access file from: ", file, ". Make sure you provide a valid path.")
    
    if (isHttp & verbose) 
        message("Remote file can take a while.\nThe full list of remote files can be found in http://aidenlab.org/data.html")
    
    ## process chromsome
    
    chromosomesChar <- NULL
    if (!is.null(chromosomes)) {
        chromosomesChar <- .chromosomes_clean(chromosomes)
    }
    
    ## process pairs
    pairClean <- NULL
    if (!is.null(pairs)) {
        pairClean <- .parse_pair(pairs)
        if (is.logical(pairClean)) {
            stop("check input pair! ", paste(pairs, collapse = ", "))
        }
    } else {
        pairClean <- .chrosomes_to_pair(chromosomesChar)
    }
    if(verbose) message("Use pairs: ", paste(pairClean, collapse = ", "))
    
    unit = match.arg(unit)
    
    if (length(resolution) > 1) {
        stop("You provide resolution: ", paste0(resolution, collapse = ","), ". Please change the resolution to a number.\n", 
            "From resolution (", unit, "): ", paste0(defResolution[[unit]], collapse = ","))
    }
    
    hicInfo <- readJuicerInformation(file)
    
    {
        unitResolution <- hicInfo[["resolution"]]
        
        stopifnot(unit %in% names(unitResolution))
        stopifnot(resolution %in% unitResolution[[unit]])
        
        chrAll <- hicInfo[["chromosomeSizes"]][["chromosome"]]
        .check_chr(pairClean, chrAll)
    }
    
    ans <- hicDataSimuFromFile(file, isHttp, pairClean, unit, resolution, seqDepth, countScale, noiseRate, 
        neighborZeroRate)
    
    ans$information <- hicInfo
    
    return(ans)
}

.dfToList <- function(contactsDf) {
    colnames(contactsDf) = c("chr1", "x", "chr2", "y", "counts")
    pairs <- paste(contactsDf$chr1, contactsDf$chr2, sep = "_")
    return(lapply(split(contactsDf[, c("x", "y", "counts")], pairs), as.matrix))
}

